### ISSUE TITLE
New warning about function spacing since version 3.12.0

### ISSUE THREAD

**Issue by ahtux** [PHPCSStandards/PHP_CodeSniffer#945]
## Describe the bug
Since version 3.12.0 PHP CodeSniffer complains about missing blank lines before a function.
You can see the minimalized code example below. There are clearly blank lines before the function.
Additionally it is the first function and due to the rule it also shouldn't complain.
[Also see discussion](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions/938)

### Code sample
```php
<?php
/**
 * Description of file
 *
 * @package foo
 */


/**
 * Description of function
 */
function foo($x)
{
}
```

### Custom ruleset
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="PHPCS rule set">
    <rule ref="Squiz.WhiteSpace.FunctionSpacing">
        <properties>
            <property name="spacingBeforeFirst" value="0" />
            <property name="spacingAfterLast" value="0" />
        </properties>
    </rule>
</ruleset>
```

### To reproduce
1. Create a file called `test.php` with the code sample above...
2. Run `phpcs test.php ...`
3. See error message displayed
```text
foo.php:12:1: error - Expected 2 blank lines before function; 0 found
```

## Expected behavior
With the code example CodeSniffer should not complain about missing blank lines.

## Versions (please complete the following information)

|                                       |                                                                                                           |
|--------------------------|----------------------------------------------------------------------  |
| Operating System           | [Linux Manjaro]                                                       |
| PHP version                    | [8.3]                                                                                     |
| PHP_CodeSniffer version | [3.12.0]                                                                            |
| Standard                         | [Squiz]                                                      |
| Install type                      | [PHAR] |

## Additional context
This behaviour ist since version 3.12.0!
With version 3.11.3 this was not an issue.

## Please confirm
- [X] I have searched the issue list and am not opening a duplicate issue.
- [X] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [X] I confirm that this bug is a bug in PHP_CodeSniffer and not in one of the external standards.
- [ ] I have verified the issue still exists in the `master` branch of PHP_CodeSniffer.

**Comment by jrfnl** [comment]
Thanks for reporting @ahtux ! That definitely is a regression. PR #946 should fix this. Testing appreciated.

**Comment by ahtux** [comment]
I can confirm that it is solved as reported...
but...
I've modified the code a bit and believe that there are still two wrong cases:

First:
```php
<?php
/**
 */


function foo()
{
}
```
Second:
```php
<?php
/**
 */
function foo()
{
}
```

In both cases PHPCodeSniffer complains about missing blank lines...

**Comment by jrfnl** [comment]
@ahtux I think these last two are based on a misunderstanding of the `spacingBeforeFirst` and `spacingAfterLast` properties.

As per the documentation (emphasis mine):

> The `spacingBeforeFirst` property is used to determine how many blank lines are required before a function when it is the first block of code inside a statement group or control structure. **This property is ignored when the function is outside one of these scopes**, or if the function is preceded by member vars. If this property has not been set, the sniff will use whatever value has been set for the `spacing` property.
>
> The `spacingAfterLast` property is used to determine how many blank lines are required after a function when it is the last block of code inside a statement group or control structure. **This property is ignored when the function is outside one of these scopes**, or if any member vars are placed after the function. If this property has not been set, the sniff will use whatever value has been set for the `spacing` property.

Ref: https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties#squizwhitespacefunctionspacing

In other words, the sniff throwing errors in those last two cases looks correct to me.
